src/extensions/arithmetic/int_ops.rs

```diff
@@ -0,0 +1,517 @@
+//! Basic integer operations.
+
+use itertools::Itertools;
+use smol_str::SmolStr;
+
+use super::super::logic::bool_type;
+use super::int_types::{get_width, int_type, INT_PARAM};
+use crate::{
+    resource::{ResourceSet, SignatureError},
+    types::{type_param::TypeArg, HashableType, SimpleRow, SimpleType, TypeRow},
+    Resource,
+};
+
+/// The resource identifier.
+pub const fn resource_id() -> SmolStr {
+    SmolStr::new_inline("arithmetic.int")
+}
+
+fn iwiden_sig(
+    arg_values: &[TypeArg],
+) -> Result<(SimpleRow, SimpleRow, ResourceSet), SignatureError> {
+    let (arg0, arg1) = arg_values.iter().collect_tuple().unwrap();
+    let m: u8 = get_width(arg0)?;
+    let n: u8 = get_width(arg1)?;
+    if m > n {
+        return Err(SignatureError::InvalidTypeArgs());
+    }
+    Ok((
+        vec![int_type(m)].into(),
+        vec![int_type(n)].into(),
+        ResourceSet::default(),
+    ))
+}
+
+fn inarrow_sig(
+    arg_values: &[TypeArg],
+) -> Result<(SimpleRow, SimpleRow, ResourceSet), SignatureError> {
+    let (arg0, arg1) = arg_values.iter().collect_tuple().unwrap();
+    let m: u8 = get_width(arg0)?;
+    let n: u8 = get_width(arg1)?;
+    if m < n {
+        return Err(SignatureError::InvalidTypeArgs());
+    }
+    Ok((
+        vec![int_type(m)].into(),
+        vec![SimpleType::new_sum(vec![
+            int_type(n),
+            HashableType::OpError.into(),
+        ])]
+        .into(),
+        ResourceSet::default(),
+    ))
+}
+
+fn itob_sig(
+    _arg_values: &[TypeArg],
+) -> Result<(SimpleRow, SimpleRow, ResourceSet), SignatureError> {
+    Ok((
+        vec![int_type(1)].into(),
```

Comment:
could this not be generic over width?

Reply:
not sure what the WASM parallel is

Reply:
WASM doesn't have a dedicated boolean type, it just uses ints. For converting bools to arbitrary-width ints we can compose this with `i_widen_u` or `iwiden_s`.